My linked Stack

koder95 opened the pull request.

koder95 committed:
My linked Stack

koder95 committed:
Ehh...

mateacademy-ai-mentor commented on src/main/java/core/stack/MateStack.java:
```diff
@@ -1,19 +1,44 @@
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static class Node<T> {
```
Inner static Node reuses the generic name `T`, which can be confusing since the outer class is `MateStack<T>`. Consider renaming the inner generic type parameter (for example `Node<E>`) or making the Node class non-static so it can use the outer `T` implicitly. This improves readability and avoids shadowing the outer generic type.

mateacademy-ai-mentor commented on src/main/java/core/stack/MateStack.java:
```diff
@@ -1,19 +1,44 @@
+    private int size = 0;
 
+    public void push(T value) {
+        Node<T> top = new Node<>(value);
```
Local variable shadows the instance field `top`. You declare `Node<T> top = new Node<>(value);` which hides the class field `top`. Rename this local variable (e.g., `node` or `newTop`) to make the code clearer and avoid confusion.

mateacademy-ai-mentor commented on src/main/java/core/stack/MateStack.java:
```diff
@@ -1,19 +1,44 @@
     public T peek() {
-        return null;
+        if (top == null) {
+            throw new EmptyStackException();
```
`peek()` correctly checks for an empty stack and throws `EmptyStackException()` when `top` is null. This satisfies the requirement to throw `EmptyStackException` for element-access operations on an empty stack—good.

mateacademy-ai-mentor commented on src/main/java/core/stack/MateStack.java:
```diff
@@ -1,19 +1,44 @@
+        Node<T> currentTop = top;
+        T value = peek();
+        top = top.prev;
+        currentTop.prev = null;
```
Severing the popped node's link with `currentTop.prev = null;` is good practice to avoid retaining references and help garbage collection. Keep this line as it prevents memory leaks in the linked-list implementation.

mateacademy-ai-mentor commented on src/main/java/core/stack/MateStack.java:
```diff
@@ -1,19 +1,44 @@
+        return value;
     }
 
     public int size() {
```
`size()` is implemented and returns the maintained `size` counter — correct. Consider adding `isEmpty()` (`return size == 0;`) and `clear()` (set `top = null; size = 0;`, and optionally sever links) if you want a more complete Stack API or if tests expect these methods.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

**Decision:** Approve — core requirements are met and functionality is correct.

Your `MateStack<T>` correctly implements a linked-list-based stack with O(1) `push`, `pop`, and `peek`, throws `EmptyStackException` on empty access, maintains `size`, and severs links on `pop` to aid GC. Nice work on those essentials.

Minor, non-blocking improvements:
- Avoid shadowing the field `top` in `push` (around line 19). Rename the local variable to something like `newNode` and use `this.top` when assigning for clarity.
- The inner `Node<T>` reuses `T`, shadowing the outer type parameter. Rename it (e.g., `private static class Node<E>`) or make `Node` non-static to use the outer `T` implicitly for better readability.
- Optional API additions: `isEmpty()` (return `size == 0`) and `clear()` (set `top = null; size = 0;` and optionally sever links in a loop) can make the class more complete if tests/consumers expect them.

Overall, the implementation should pass standard tests. Keep the good practices you already have (empty checks in `peek`, reference nulling in `pop`).

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

koder95 committed:
Refactor pop method variable declaration order

Moved the declaration of the 'value' variable to the beginning of the pop() method for improved readability and consistency.

MateuszRostek commented on src/main/java/core/stack/MateStack.java:
```diff
@@ -1,19 +1,44 @@
+    private int size = 0;
 
+    public void push(T value) {
+        Node<T> top = new Node<>(value);
+        top.prev = this.top;
+        this.top = top;
+        size++;
```
I would suggest creating and using constructor with all fields, something like this:

```suggestion
        Node<T> first = head;
        head = new Node<>(first, value);
        size++;
```

MateuszRostek reviewed (approved):
Good job! Check one small advice from my side + comments from ai-mentor
